platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -149,25 +171,23 @@ perhaps the node is so far behind that none of its peers are able to supply any
 to operate, and must reconnect. Gossip will make a call through the Consensus module interface to notify Execution that
 gossip is "down". Execution will then initiate reconnect.
 
-Fundamentally, Execution is responsible for reconnect, but it cannot differentiate from a quiescent network or a
+Fundamentally, Execution is responsible for reconnect, but it cannot differentiate between a quiescent network or a
 behind node just by looking at the lack of rounds coming from Consensus. So Consensus **must** tell Execution if it
 knows that it cannot proceed in creating rounds because the node itself is bad, vs. a quiescent network which is
 able to create rounds (if there were any transactions).
 
 #### Roster Changes
 
 At runtime, it is possible that the roster will change dynamically (as happens with the dynamic address book feature).
-Roster changes at the gossip level influences which peers the module will work with. Because different peers may be
-at different points in consensus time, it may be that Alice has a newer roster than this node does, or Bob may have an
-older roster. Bob still needs to be able to gossip, even if he is behind. So each node will need to be able to honor
-connections from all peers in all rosters associated with rounds that have not expired.
-
-As with all other modules using rosters, Gossip must have a deterministic understanding of which roster applies to
-which round. It will receive this information from Hashgraph in the form of round metadata.
+Roster changes at the gossip level may influence which peers the module will work with. As with all other modules using
+rosters, Gossip may need a deterministic understanding of which roster applies to which round. It will receive this
+information from Hashgraph in the form of round metadata.
```

Comment:
As we discussed offline, when a roster change is signaled from Execution to Consensus, it is done by making a call to the event intake system. Event intake will include the round information in the next outgoing event. The Hashgraph module will observe events with "new round info" and track which nodes have reported this new round. When sufficient nodes by stake weight have claimed the same new round, then it will become effective.

When it becomes effective, it will be transmitted through the round information, which among other things, contains the address book effective for that round.

Reply:
I recall discussing this strategy. From my perspective, this is one of those things where the cure is worse than the disease. Making consensus maintain a state outside of "the state" is a recipe for some unpleasant complexities.

- consensus state needs to be written to disk during state snapshots
- consensus state needs to be hashed and signed
- consensus state needs to be transmitted during a reconnect
- consensus state might ISS, and so will need its own ISS detection
- individual events are not guaranteed to reach consensus. So it is possible that a requested address book change may never take effect. Stale event detector is insufficient to solve this problem alone -- only way to guarantee eventual re-submission in the presence of restarts/reconnects is to build out the same system I was proposing for  guaranteed state signing. (Note that this system would have to be built on top of the consensus state, not the regular state.)

You said a while back that bad APIs are often the sign that module boundaries haven't been drawn well. This strikes me as one of those situations. Removing the cyclical data flow between consensus and execution looks nice on paper, but has some major problems in practice. Implementing dynamic address book the way I proposed was a very big change. Implementing it this way will be three or four times harder.